video: drivers misc fixes

ngphibang opened the pull request:
- Some misc fixes in individual video drivers
- Moving the format pitch (bytesperline) which relating to image size from application into bridge drivers. The reason for that is the bridge (e.g., DMA, ISP) drivers actually handle the memory and know exactly the memory layout constraints (need alignment, padding, etc. or not). Application does not know this, it justs set the desired pixel format and resolution and must always read back this field to see what the driver actually sets. Also, drop format pitch setting in sensor drivers as this is not needed (and the sensor does not know this information neither). Some examples:

https://elixir.bootlin.com/linux/v6.11.1/source/drivers/media/platform/qcom/venus/vdec.c#L231
https://elixir.bootlin.com/linux/v6.11.1/source/drivers/media/platform/ti/omap3isp/ispvideo.c#L143
https://elixir.bootlin.com/linux/v6.11.1/source/drivers/staging/media/sunxi/sun6i-isp/sun6i_isp_capture.c#L404

- Decoupling get / set_fmt() in some bridge drivers

0xFarahFl committed:
drivers: video: mt9m114: Make the driver multi-instance

The mt9m114 camera driver used to be single-instance.
Improve it to multi-instance.

Signed-off-by: Farah Fliss <farah.fliss@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

0xFarahFl committed:
drivers: video: mt9m114: Fix coding style

Fix coding style in a variable naming.

Signed-off-by: Farah Fliss <farah.fliss@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: ov5640: Drop cur_pixrate internal variable

Update the control value directly. No need for an internal variable.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: mipi_csi2rx: Fix type range related to pixel rate

Fix some type range related to pixel rate which can cause overflow.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: ov5640: Fix HUE register write

SDE_CTRL8_REG's value must be modified using modify_register.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: ov5640: Fix brightness control register

Fix the sign register for brightness control

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: ov5640: Fix constrast value sign

Fix sign's register for constrast value.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: mipi_csi2rx: Explicitly set init priority

The MIPI CSI-2 Rx needs to be initialized after the camera sensor which
is generally initialized with CONFIG_VIDEO_INIT_PRIORITY.

This is currently true "by chance" due to the order the linker links the
object files. This linker order is not easily controlled, so use an
explicit priority value to ensure this requirement.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
@josuah : The video emul imager exposes 3 controls : VIDEO_CID_EXPOSURE, VIDEO_CID_GAIN and VIDEO_CID_TEST_PATTERN. However, it does nothing specific to each control type, just a "fake" register write (?). I know this is just for testing purpose but even if this is just an emulation, we need to emul something "real". If not, it will be very difficult to [rework the video control framework](https://github.com/zephyrproject-rtos/zephyr/pull/82158) as each control type needs a specific implementation. I am going to remove these controls in the driver and just keep 1 control of type "vendor-specific" as an emulated read/write. Is it ok for you to do that ?

josuah commented on drivers/video/ov5640.c:
```diff
@@ -997,25 +998,35 @@ static int ov5640_set_ctrl_contrast(const struct device *dev, int value)
+{
+	const struct ov5640_config *cfg = dev->config;
+	struct ov5640_data *data = dev->data;
+	int ret;
```
`ret` does not appear to be used, maybe CI will complain about it.

ngphibang replied:
Thank you for the patch (it get merged in your PR). I will comback to this with some more fixes on video flush

josuah replied:
Very glad to see all these sparse fixes coming in!

ngphibang commented:
- drivers: video: emul_rx: Explicitly set init priority

josuah commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -216,21 +215,20 @@ static int mipi_csi2rx_get_frmival(const struct device *dev, enum video_endpoint
-	       cur_fmival->numerator * cur_pixel_rate /
-	       (mipi_csi2rx_cal_frame_size(fie_format) * fie_frmival->numerator *
-		cur_fmival->denominator);
+	return (mipi_csi2rx_cal_frame_size(cur_format) / mipi_csi2rx_cal_frame_size(fie_format) *
+		fie_frmival->denominator / fie_frmival->numerator * cur_fmival->numerator /
+		cur_fmival->denominator * cur_pixel_rate);
```
Though, depending on what this is used for, this could be a more correct implementation than a possible integer overflow on large formats.

ngphibang replied:
You are right ! I will rethink of it.

ngphibang replied:
Fixed

ngphibang commented:
@dleach02 I need to solve the uint64_t and uint32_t issue and to add a fix to the flushing part of the CSI driver (which is important to show the need of a unified buffer management framework). I will try to do this at the end of the week.
Added DNM tag but now also convert this into draft.

ngphibang committed:
drivers: video: Compute bits per pixel according to format

Compute bits per pixel according to the pixel format instead of
hardcoding it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
The patch to rework the CSI driver's flushing part is done but it depends on some change in the NXP's HAL SDK (which often takes long) so I will push it later in another PR.

avolmat-st commented on doc/releases/release-notes-2.1.rst:
```diff
@@ -300,7 +300,7 @@ Drivers and Sensors
 
 * Video
 
-  * Added MCUX CSI and Aptina MT9M114 drivers
```
I suppose release-notes-2.1 shouldn't be modified now since it is an old release.

avolmat-st commented on dts/bindings/video/aptina,mt9m114.yaml:
```diff
@@ -5,7 +5,7 @@
 
 description: MT9M114 CMOS video sensor
 
-compatible: "aptina,mt9m114"
```
if compatible is changing, maybe an entry within the migration guide would be necessary

kartben replied:
It is required, yes. But I would actually recommend not changing the compatible unless there is a really strong case for it.
Would be great to avoid breaking downstream users by renaming compatibles every time a silicon company acquires another, basically :)

avolmat-st reviewed (changes requested):
Commit message: Fix some type type range related to pixel rate which can cause overflow:

duplicated "type"

ngphibang committed:
drivers: video: stm32-dcmi: Drop video_stm32_dcmi_is_fmt_valid

Drop video_stm32_dcmi_is_fmt_valid() as it is not needed. In this
function, (i) checking against a format based on another utility
function video_bits_per_pixel() is not robust, this check is done in
the sensor driver, (ii) checking against the heap size is not
appropriate because this should be done when allocating buffers, not
in get/set format.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Move format pitch setting to bridge drivers

The format pitch (bytesperline) field is typically set by the bridge
drivers, i.e. DMA, ISP drivers who actually handle the memory as they
know exactly the memory layout constraints.

Application just set the pixel format and resolution and must always
read back this field to see what the driver actually sets (to allocate
buffers for example).

Also, drop format pitch setting in sensor drivers as this is not needed.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: mcux_csi: Don't set_format() in get_format()

Do not set_format() when doing get_format(). This design seems initially
to simplify the sample (just get_format() and everything works out of the
box) but it makes thing incomprehensive and error prone.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: mcux_smartdma: Don't set_format() in get_format()

Do not set_format() when doing get_format(). This design seems initially
to simplify the sample (just get_format() and everything works out of the
box) but it makes thing incomprehensive and error prone.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
Forced push:

- Dropped the commit that fix company name (aptina vs onsemi)
Add some commits for

- Moving the format pitch (bytesperline) which relating to image size from application into bridge drivers. The reason for that is the bridge (e.g., DMA, ISP) drivers actually handle the memory and know exactly the memory layout constraints (need alignment, padding, etc. or not). Application does not know this, it justs set the desired pixel format and resolution and must always read back this field to see what the driver actually sets. Also, drop format pitch setting in sensor drivers as this is not needed. Some examples:

https://elixir.bootlin.com/linux/v6.11.1/source/drivers/media/platform/qcom/venus/vdec.c#L231
https://elixir.bootlin.com/linux/v6.11.1/source/drivers/media/platform/ti/omap3isp/ispvideo.c#L143
https://elixir.bootlin.com/linux/v6.11.1/source/drivers/staging/media/sunxi/sun6i-isp/sun6i_isp_capture.c#L404

-
- Decoupling get / set_fmt() in some bridge drivers

avolmat-st commented:
Thanks for the update. For further updates, could you avoid updating the serie while rebasing at the same time ? Except if I haven't found the right way to do that, I cannot check the difference between the previous push and the last push so it becomes in such case necessary to do again the whole review of all commits.  Could it be 2 separate pushes in such cases, one that do the rebase (hence no code change really except rebasing) and a 2nd one which updates the serie itself (could be the opposite as well).

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -186,17 +186,6 @@ static int stm32_dcmi_enable_clock(const struct device *dev)
 	return clock_control_on(dcmi_clock, (clock_control_subsys_t *)&config->pclken);
 }
 
-static inline int video_stm32_dcmi_is_fmt_valid(uint32_t pixelformat, uint32_t pitch,
```
I agree that the heap size is not correct here, checking of buffer is actually already done within the enqueue so that should be enough. Considering that the DCMI doesn't perform any transformation on the data (currently at least) there is no point at checking currently that here via dcmi_is_fmt_valid.

ngphibang replied:
To be sure I understand well your comment, so does it mean it's ok as I did or you want to keep the check with video_bits_per_pixel() ?

(Forrmat check is done at the sensor but it is not a problem if we do it here first, I will put it back if required)

avolmat-st replied:
This patch is alright. With the current DCMI driver there is no reason to keep it as the sensor with check the format. I am thinking about adding some more features, such as compose / crop, will see at this moment if this becomes necessary. I need to first push the crop / compose API first since I also need it for the DCMIPP. Will prepare that rapidly.
Anyway for this PR this code is ok for me.

ngphibang commented:
> Thanks for the update. For further updates, could you avoid updating the serie while rebasing at the same time

Yes, I'll do it next times. Thanks !

avolmat-st commented on drivers/video/video_mcux_smartdma.c:
```diff
@@ -294,16 +294,7 @@ static int nxp_video_sdma_get_format(const struct device *dev,
 	    (fmt->width != fmts[0].width_min) ||
 	    (fmt->height != fmts[0].height_min) ||
 	    (fmt->pitch != fmts[0].width_min * 2)) {
```
the pitch check doesn't really make sense now I think.

ngphibang replied:
Yes, I missed this. Done in `set_format()` too. Thank you !

avolmat-st commented:
Seems to me there is again a rebase with updates of the code at the same time, leading to
        Showing 434 changed files with 9,796 additions and 29,815 deletions.
In such situation it is not possible to perform a Compare review that would have been most probably very easy.

Please ensure to only perform updates on top of the previously pushed branch.

cf 3rd point of https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

ngphibang commented:
> Seems to me there is again a rebase with updates of the code at the same time

Oh, sorry ... I did it as per usual to always test the PR on the latest main to see if there is a problem and pushed it unconsciously. Will pay attention. Sorry again !

avolmat-st commented on drivers/video/ov5640.c:
```diff
@@ -996,9 +996,13 @@ static int ov5640_set_ctrl_hue(const struct device *dev, int value)
+					  {SDE_CTRL2_REG, abs(sin_coef) & 0xFF}};
+
+	ret = ov5640_modify_reg(&cfg->i2c, SDE_CTRL8_REG, 0x7F, sign);
+	if (ret) {
```
Recently we had a discussion with @josuah in which we discussed about whether or not we should enforce MISRA style. Normally this is apparently a Zephyr requirement but is not enforced so hard to decide actually. In doubt I update by ST-MIPID02 / GC2145 PR few days ago for all the code I modified (I didn't modify the whole driver when the driver already existed) to if in such similar situation if (ret < 0). I tend to actually not being a big fan of it but that was more in preparation of this style being enforced.
I honestly don't really know if we should enforce it now or not. An opinion ?

ngphibang replied:
> Recently we had a discussion with @josuah in which we discussed about whether or not we should enforce MISRA style.

To me, it's the same. Without MISRA rule, it's a bit shorten. Personally, I am not fan of it as well, e.g. when it comes to `if else` the style is a bit ugly to me:

https://github.com/zephyrproject-rtos/hal_nxp/blob/a961ce3ba3fa3a4f5c6d84918b1d2262197f46d3/mcux/mcux-sdk-ng/drivers/csi/fsl_csi.c#L864-#L879

But well, if it's enforced by Zephyr one day, we need to follow. Currently, to put it simple, I will change this in the next push ... personnaly, I tend to avoid any discussion / debate about things which are less technical :-).

josuah replied:
@pdgendt mentioned:

> It's a required rule, but it's not enforced (yet) in CI

There is a doc page that tells the roadmap of for different level of coding style enforcement (including MISRA, and listing of areas) but I cannot find it again. :sweat_smile:

avolmat-st reviewed (commented):
I am ok with the PR, with only the point about MISRA or not ;) just have to be decided maybe. Also the PR will have to be rebased since now the endpoint removal as been merged as far as I know.

ngphibang commented:
@avolmat-st : To make sure I will push the branch correctly as you want this time:

- I rebased the branch on top of the lastest main and resolved the conflict when rebasing, then I push, right ? There are still many unrelated changes coming from main together with the conflict resolving though, is it ok like that ?

ngphibang commented:
Simply rebased and fixed conflict because https://github.com/zephyrproject-rtos/zephyr/pull/87070 was merged

ngphibang commented:
Forced push:

- Replaced `if (ret)` by `if (ret < 0)` to be compliant with MISRA rule.

avolmat-st commented:
> @avolmat-st : To make sure I will push the branch correctly this time:
> 
> * I rebased the branch on top of the lastest main and resolved the conflict when rebasing, then I push, right ? There are still many unrelated changes coming from main together with the conflict resolving though, is it ok like that ?

Sorry for the late reply. Yes thanks a lot !!

josuah commented on drivers/video/video_emul_rx.c:
```diff
@@ -67,7 +67,9 @@ static int emul_rx_set_fmt(const struct device *const dev, struct video_format *
 	}
 
 	/* Cache the format selected locally to use it for getting the size of the buffer  */
+	fmt->pitch = fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE;
```
This is a small semantic change and I proposed a migration-guide entry here to avoid getting in the way of this PR:
- https://github.com/zephyrproject-rtos/zephyr/pull/90110

This is not a breaking change, and a migration guide entry is only a hint for how to improve applications.

erwango commented:
@mmahadevan108, @dleach02 PTL
